air-script/tests/bitwise/bitwise.masm

```diff
@@ -1,38 +1,37 @@
 proc.cache_z_exp
-    # Load trace length
-    mem_load.4294903306
-    # => [trace_len, ...]
-    # Push initial num_of_cycles
-    push.1
-    # => [num_of_cycles, trace_len, ...]
-    # Load Z
     padw mem_loadw.4294903304 drop drop
-    # => [z_1, z_0, num_of_cycles, trace_len, ...]
-    # Compute exponentiations based on the number of cycles for a period of 8
-    dup.3 div.8
-    # => [num_of_cycles', z_1, z_0, num_of_cycles, trace_len, ...]
-    # Update next num_of_cycles and compute number of iterations
-    movup.3 dup.1 movdn.4 div
-    # => [i, z_1, z_0, num_of_cycles', trace_len, ...]
-    # Exponentiate the existing `z**num_of_cycles` an additional `i` times
-    dup.0 neq.1
+    # => [z_1, z_0, ...]
+    # Number of exponentiations for a period of length 8
+    mem_load.4294903307 neg add.3
+    # => [count, z_1, z_0, ...]
+    # Exponentiate z
+    dup.0 neq.0
     while.true
         movdn.2 dup.1 dup.1 ext2mul
-        # => [z_1^2, z_0^2, i, num_of_cycles', trace_len, ...]
-        movup.2 div.2 dup.0 neq.1
-        # => [b, i+1, z_1^2, z_0^2, num_of_cycles', trace_len, ...]
+        # => [z_1^2, z_0^2, i, ...]
+        movup.2 add.1 dup.0 neq.0
+        # => [b, i+1, z_1^2, z_0^2, ...]
     end # END while
-    drop push.0 push.0 mem_storew.500000100 # Save the exponentiation of z for column 0
-    mem_storew.500000101 # Save the exponentiation of z for column 1
-    # => [0, 0, z_1^2, z_0^2, num_of_cycles', trace_len, ...]
+    push.0 mem_storew.500000100 # z^exp for period 8
+    # => [0, 0, z_1^2, z_0^2, ...]
     drop drop
-    # => [z_1^2, z_0^2, num_of_cycles', trace_len, ...]
+    # Exponentiate z 8 times, until trace_len
+    push.18446744069414584313 dup.0 neq.0
+    while.true
+        movdn.2 dup.1 dup.1 ext2mul
+        # => [z_1^2, z_0^2, i, ...]
+        movup.2 add.1 dup.0 neq.0
+        # => [b, i+1, z_1^2, z_0^2, ...]
+    end # END while
+    push.0 mem_storew.500000101 # z^exp for trace_len
+    # => [0, 0, z_1^2, z_0^2, ...]
+    drop drop
+    # => [z_1^2, z_0^2, num_of_cycles', -log2(trace_len), ...]
     # Clean stack
     dropw
 end # END PROC cache_z_exp
 
 proc.cache_periodic_polys
-    exec.cache_z_exp
     # periodic column 0
     padw mem_loadw.500000100 drop drop
     # => [z_exp_1, z_exp_0, ...]
```

Comment:
nit: if one of the elements is in the base field, then `ext2mul` can be substituted by two base field multiplications. This is basically the `<Felt as ExtensibleField<2>>::mul_base` method

Reply:
I'm not sure I follow:

1. Are you suggesting something like `[(q1, q0), b, ...]` where `(q1, q0)` is an quadratic extension, and `b` a base field, that we should use `dup.2 mul swap movup.2 mul swap` instead of `ext2mul`? That would be `6` cycles against `5` from `movup.2 push.0 ext2mul`
2. This procedure should be using only quadratic elements, it corresponds only to the [polynomial evaluation](https://github.com/facebook/winterfell/blob/main/verifier/src/evaluator.rs#L34). Unless I made a mistake that should be only extension field elements, no?

Reply:
I thought that avoiding `ext2mul` will be better but as you showed it is worse than your solution. Thank you for the clarification.
Yes, the evaluation point is an element of the quadratic extension field.